File: go/internal/verify/keys.go
package verify

import (
	"encoding/base64"
	"fmt"
)

// NormalizeEd25519PublicKey accepts a raw 32-byte key ([]byte) or a base64
// string encoding those 32 bytes. Rejects PEM SPKI, hex, and other formats.
// The caller wraps the returned error as a CertificateError with
// ReasonInvalidSignature so a malformed key still surfaces as a typed
// SDK error.
func NormalizeEd25519PublicKey(input any) ([]byte, error) {
	var raw []byte
	switch v := input.(type) {
	case []byte:
		raw = v
	case string:
		b, err := base64.StdEncoding.DecodeString(v)
		if err != nil {
			return nil, fmt.Errorf("ed25519 public key base64 decode failed: %w", err)
		}
		raw = b
	case nil:
		return nil, fmt.Errorf("ed25519 public key is nil")
	default:
		return nil, fmt.Errorf("ed25519 public key must be []byte or base64 string, got %T", input)
	}
	if len(raw) != 32 {
		return nil, fmt.Errorf("ed25519 public key must be 32 bytes, got %d", len(raw))
	}
	return raw, nil
}

File: python/src/theveil/verify_certificate/keys.py
"""Ed25519 public-key input normalization.

Accepts raw 32-byte keys or base64 strings. Rejects PEM SPKI, hex, and
other formats — the SDK's contract is raw-bytes-or-base64 only.
"""

from __future__ import annotations

import base64

__all__ = ["normalize_ed25519_public_key"]


def normalize_ed25519_public_key(input_: bytes | str) -> bytes:
    """Normalize a witness public-key input to raw 32 bytes.

    Raises:
        TypeError: on unsupported input type, wrong length, or invalid
            base64. The caller (:func:`verify_certificate`) wraps these
            as ``TheVeilCertificateError(reason="invalid_signature")`` so
            a malformed key still surfaces as a typed SDK error.
    """

    if isinstance(input_, bytes):
        raw = input_
    elif isinstance(input_, bytearray):
        raw = bytes(input_)
    elif isinstance(input_, str):
        try:
            raw = base64.b64decode(input_, validate=True)
        except (ValueError, base64.binascii.Error) as exc:  # type: ignore[attr-defined]
            raise TypeError(f"Ed25519 public key base64 decode failed: {exc}") from exc
    else:
        raise TypeError(
            "Ed25519 public key must be bytes or base64 string, got "
            f"{type(input_).__name__}"
        )
    if len(raw) != 32:
        raise TypeError(f"Ed25519 public key must be 32 bytes, got {len(raw)}")
    return raw
